src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/ConversionsBase.cs

```diff
@@ -1266,6 +1265,15 @@ private static LambdaConversionResult IsAnonymousFunctionCompatibleWithDelegate(
                 return LambdaConversionResult.BadTargetType;
             }
 
+            if (anonymousFunction.HasExplicitReturnType(out var refKind, out var returnType))
+            {
+                if (invokeMethod.RefKind != refKind ||
+                    !invokeMethod.ReturnType.Equals(returnType.Type, TypeCompareKind.AllIgnoreOptions))
```

Comment:
Are we sure this is the rule we want? Does this fall out of existing spec rules, or needs to be spec'ed?

On one hand, `Func<int> f = localFunctionReturningShort;` an error (with `short localFunctionReturningShort()`).  So `Func<int> f = short () => 1;` makes sense to error as well.
On the other hand, I think `Func<object, int> f = (o) => 1;` involves a conversion from expression which gives `o` its `object` type, so why wouldn't that conversion also handle `Func<object, int> f = short (o) => 1`?

Reply:
Also, is `AllIgnoreOptions` the right option? Can we include tests for types differing by small differences (object vs. dynamic, nullability, custom modifiers)?

Reply:
> Are we sure this is the rule we want? Does this fall out of existing spec rules, or needs to be spec'ed?

This should be confirmed with LDM, but I believe it corresponds to the behavior for parameters. For instance, `Action<string> a = (object o) => { };` results in a conversion error.

> Can we include tests for types differing by small differences (object vs. dynamic, nullability, custom modifiers)?

There are tests for `object` and `dynamic`, tuple element names, `nint` and `IntPtr`, and nullable differences.

Reply:
You're correct, this is the behavior for parameters: [https://github.com/dotnet/csharplang/blob/main/spec/conversions.md#anonymous-function-conversions](https://github.com/dotnet/csharplang/blob/main/spec/conversions.md#anonymous-function-conversions)
"If F has an explicitly typed parameter list, each parameter in D has the same type and modifiers as the corresponding parameter in F."

We probably need to add a similar rule to the spec for the return type.

Reply:
These things should just fall out. Variance only allows for implicit reference conversions, nothing else, because they need to be representation-preserving.

Reply:
> Is there a way to get into a situation with a custom modifiers difference as well?

Added tests for custom modifiers.